tests/5.1/target/test_target_has_device_addr.c

```diff
@@ -0,0 +1,96 @@
+//===--- test_target_has_device_addr.c --------------------------------------===//
+//
+// OpenMP API Version 5.1 Nov 2020
+//
+// This test verifies the 'has_device_addr' feature added to the target construct.
+// We tested this by mapping a scalar & array to the device (using both target map
+// & omp_target_alloc) and ensuring that the address does not change when using
+// has_device_addr on another target region.
+//
+////===----------------------------------------------------------------------===//
+#include <assert.h>
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1000
+
+//test by mapping to device use 'target map' construct
+int test_target_has_device_addr() {
+  OMPVV_INFOMSG("test_target_has_device_addr");
+  int errors = 0;
+  int x = 10;
+  int arr[N];
+  for(int i=0;i<N;i++){
+    arr[i] = i;
+  }
+  int *first_scalar_device_addr;
+  int *first_arr_device_addr;
+  #pragma omp target enter data map(to: x, arr)
+  #pragma omp target map(from: first_scalar_device_addr, first_arr_device_addr) map(to: x, arr)
+  {
+    first_scalar_device_addr = &x;
+    first_arr_device_addr = &arr[0];
+  }
+  int *data_x, *data_arr;
+  int *second_scalar_device_addr, *second_arr_device_addr;
+  //check addresses are same on both target data region & device region
+  #pragma omp target data use_device_addr(x,arr)
+  {
+    data_x = &x; data_arr = &arr[0];
+    #pragma omp target map(from:second_scalar_device_addr, second_arr_device_addr) has_device_addr(x, arr) 
+    {
+      second_scalar_device_addr = &x;
+      second_arr_device_addr = &arr[0];
+    }
+  }
+  #pragma omp target exit data map(release: x, arr)
+  OMPVV_TEST_AND_SET(errors, first_scalar_device_addr != data_x);
+  OMPVV_TEST_AND_SET(errors, first_arr_device_addr != data_arr);
+  OMPVV_TEST_AND_SET(errors, first_scalar_device_addr != second_scalar_device_addr);
+  OMPVV_TEST_AND_SET(errors, first_arr_device_addr != second_arr_device_addr)
+  return errors;
+}
+
+//test by mapping to device use 'omp_target_alloc' construct
+int test_target_has_dev_addr_alloc(){
+
+  OMPVV_INFOMSG("test_target_has_device_addr_alloc");
+  int errors = 0;
+  int x = 10;
+  int arr[N];
+  for(int i=0;i<N;i++){
+    arr[i] = i;
+  }
+  int dev = omp_get_default_device();
+  int *first_scalar_device_addr = (int *)omp_target_alloc(sizeof(int), omp_get_default_device());
+  int *first_arr_device_addr = (int *)omp_target_alloc(sizeof(int) * N, omp_get_default_device());
+  #pragma omp target enter data map(to: x, arr)
+  int *data_x, *data_arr;
+  int *second_scalar_device_addr, *second_arr_device_addr;
+  //check addresses are same on both target data region & device region
+  #pragma omp target data use_device_addr(x,arr)
+  {
+    data_x = &x; data_arr = &arr[0];
```

Comment:
why do you need this ? 
For testing it is adequate to record the device address in target1 and target2 and compare the two. I think @tob2 is saying the same thing.